internal/topology/variables/cluster_variable_validation.go

```diff
@@ -151,34 +167,66 @@ func ValidateClusterVariable(value *clusterv1.ClusterVariable, definition *clust
 		return err
 	}
 
-	return validateUnknownFields(fldPath, value, variableValue, apiExtensionsSchema)
-}
-
-// validateUnknownFields validates the given variableValue for unknown fields.
-// This func returns an error if there are variable fields in variableValue that are not defined in
-// variableSchema and if x-kubernetes-preserve-unknown-fields is not set.
-func validateUnknownFields(fldPath *field.Path, clusterVariable *clusterv1.ClusterVariable, variableValue interface{}, variableSchema *apiextensions.JSONSchemaProps) field.ErrorList {
 	// Structural schema pruning does not work with scalar values,
 	// so we wrap the schema and the variable in objects.
 	// <variable-name>: <variable-value>
 	wrappedVariable := map[string]interface{}{
-		clusterVariable.Name: variableValue,
+		definition.Name: variableValue,
 	}
 	// type: object
 	// properties:
 	//   <variable-name>: <variable-schema>
 	wrappedSchema := &apiextensions.JSONSchemaProps{
 		Type: "object",
 		Properties: map[string]apiextensions.JSONSchemaProps{
-			clusterVariable.Name: *variableSchema,
+			definition.Name: *apiExtensionsSchema,
 		},
 	}
 	ss, err := structuralschema.NewStructural(wrappedSchema)
 	if err != nil {
-		return field.ErrorList{field.Invalid(fldPath, "",
-			fmt.Sprintf("failed defaulting variable %q: %v", clusterVariable.Name, err))}
+		return field.ErrorList{
+			field.InternalError(fldPath,
+				fmt.Errorf("failed to create structural schema for variable %q; ClusterClass should be checked: %v", value.Name, err))} // TODO: consider if to add ClusterClass name
+	}
+
+	if err := validateUnknownFields(fldPath, definition.Name, wrappedVariable, ss); err != nil {
+		return err
+	}
+
+	celValidator := cel.NewValidator(ss, false, celconfig.PerCallLimit)
+	// celValidation will be nil if there are no CEL validations specified in the schema
+	// under `x-kubernetes-validations`.
+	if celValidator == nil {
+		return nil
+	}
+
+	// Only extract old variable value if there are CEL validations and if the old variable is not nil.
+	var oldWrappedVariable map[string]interface{}
+
+	if oldValue != nil && oldValue.Value.Raw != nil {
+		var oldVariableValue interface{}
+
+		if err := json.Unmarshal(oldValue.Value.Raw, &oldVariableValue); err != nil {
+			return field.ErrorList{field.Invalid(fldPath.Child("value"), string(oldValue.Value.Raw),
+				fmt.Sprintf("old value of variable %q could not be parsed: %v", value.Name, err))}
+		}
+
+		oldWrappedVariable = map[string]interface{}{
+			definition.Name: oldVariableValue,
+		}
 	}
 
+	if err, _ := celValidator.Validate(ctx, fldPath, ss, wrappedVariable, oldWrappedVariable, celconfig.RuntimeCELCostBudget); err != nil {
```

Comment:
Q: I assume this code automatically figures out if deeper nested old values exist or not? (and only runs transition rules if they exist)

Reply:
Yes, I think so.